### ISSUE TITLE
broken pipe error when using ipcReceiver as Stream with tokio

### ISSUE THREAD

**Issue by another-s347** [servo/ipc-channel#221]
The child code is adopted from cross_process_embedded_senders() in test.rs.
```    let (tx1, rx1): (IpcSender<IpcMessage>, IpcReceiver<IpcMessage>) = channel().unwrap();
    let tx0 = IpcSender::connect(server0_name).unwrap();
    tx0.send(tx1).unwrap();
    let tx2: IpcSender<IpcMessage> = IpcSender::connect(server1_name).unwrap();
    tx2.send(IpcMessage { s: "handshake".to_string() }).unwrap();
    let mut s = rx1.map_err(|_|()).for_each(|m|{
        println!("receive:{:?}",&m);
        futures::future::ok(())
    });
    tokio::run(s);
    println!("shutdown");
```
The child can receive message successfully once. Then, the parent process's ipcSender.send() gives an error as this, while the child's tokio::run is not stopped:
```
thread '...' panicked at 'called `Result::unwrap()` on an `Err` value:
Io(Os { code: 32, kind: BrokenPipe, message: "Broken pipe" })', src/libcore/result.rs:999:5
```
However, a loop with ipcReceiver.recv() can just work.

Rust version: rustc 1.33.0-nightly (e2f221c75 2019-01-15)
Ipc-channel = {version="0.11.2",features=["async"]}
futures = "0.1"
tokio = "0.1.14"
os: Manjaro with Kernel 4.19.13

**Comment by dlrobertson** [comment]
Do you have a complete example? I tried adapting your code to a point where it compiles, but I wasn't able to repro your issue.

**Comment by another-s347** [comment]
This is the minimal code to show the problem: https://github.com/another-s347/PlayIpcChannel

It looks like when using with tokio, the socket get closed shortly after tokio::run for some reason. If parent's sender waits for a little, the error comes up.

**Comment by dlrobertson** [comment]
:+1: thanks for the example @another-s347. Ran your code and after some debugging realized that the the `fork` seems to exit when we return `NotReady` on `poll`. This may be related to #219. We might need to inform `tokio` when to wake, but I'm not up to date with all-things `tokio` or `futures`.

**Comment by asajeffrey** [comment]
Hopefully, #227 fixes this, but it's for futures 0.3.

**Comment by another-s347** [comment]
I update the demo for #227 , and the issue has been resolved. Thanks for your work.
